### PR TITLE
Fix false positive hash join detection

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Closes https://github.com/crate/crate/issues/16173

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java [hunk 1]
```diff
@@ -1169,4 +1170,19 @@ CONSTRAINT int_check CHECK("int_col" > 20)
...
+        execute("CREATE TABLE  t0 (c1 INT NOT NULL DEFAULT 1)");
+        execute("CREATE TABLE  t1 (c1 INTEGER)");
+        execute("INSERT INTO t0(c1) VALUES (1)");
+        execute("INSERT INTO t1(c1) VALUES (2)");
+        execute("REFRESH TABLE t0");
+        execute("REFRESH TABLE t1");
+
+        execute("SELECT * FROM t0, t1 WHERE (t1.c1 >= 1)=((t1.c1 = t1.c1)AND(t0.c1 <= t0.c1))");
+        assertThat(response).hasRows("1| 2");
+    }
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java:1187]: temporal thing, will remove it before merging as there is a unit test.

**Comment by BaurzhanSakhariev** [comment]
breaks existing test `testPossibleOnInnerContainingEqAndAnyCondition`

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java [hunk 2]
```diff
@@ -119,4 +119,16 @@ public void test_equality_expression_followed_by_case_expression() {
...
+        Symbol joinCondition = sqlExpressions.asSymbol("(t1.a >= 1) = ((t1.a = t1.a) AND (t2.b <= t2.b))");
+        assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isFalse();
+    }
+
+    @Test
+    public void test_ineqiality_operator_both_sides_have_constants() {
+        Symbol joinCondition = sqlExpressions.asSymbol("(t1.a > 1) = (t2.b < 5)");
+        assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isTrue();
+    }
 }
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:128]: Fails on master, passes if I reverthttps://github.com/crate/crate/commit/6e60b6674f3b011d5d56c257a3576b34ee77579

**Comment by BaurzhanSakhariev** [comment]
There are 2 different issues:

1. Reported issue (false positive HashJoin detection)

    Check query  `EXPLAIN SELECT * FROM t0, t1 WHERE (t1.c1 >= 1)=((t1.c1 = t1.c1)AND(t0.c1 <= t0.c1))`

     I enforced hash join in PG by enabling it and disabling all other join methods  - PG anyway falls back to NestedLoopJoin.
    (see first `EXPLAIN SELECT` in https://www.db-fiddle.com/f/3siUQViqcmmUomBHftGuJs/5)

     `NestedLoopJoin` must be used for the query ^. But It's **NOT** because of inequality. Inequality is ok here as it's just 
      part of the boolean expression (part of the left or right side of the EQ operator).

     See another queries with inequality  in the linked db-fiddle. (PG uses HashJoin for them because it's just part of the expression, we hash boolean output of the expression and compare true/false on build/probe phase.

1. False negative HashJoin detection - short circuit when encountering OR. If we have an EQ operator with boolean expressions, OR/<= doesn't matter, we should just compute an expression and build hash/probe for booleans.

    Check query  `EXPLAIN SELECT * FROM t0, t1 WHERE (t1.c1 >= 1)=(t0.c1 <= 5 OR t0.c1 <= t0.c1)`
    
    It's the last query in the db-feedle, PG can use HashJoin for it.

**Comment by BaurzhanSakhariev** [comment]
Update on issue 2 (CrateDB only falsely says that HashJoin is not possible and uses NestedLoopJoin). However, result is correct so probably not worth fixing.

Reason: When conparing booleans HashJoin is pretty mch useless, it behaves almost in the same way as NestedForLoop, in worst case it's same thing. We have a map with 2 keys (true/false) and each key miltiple collisions. If, say probing side has 1 million records, we have at least 500000 collisions for true or false key, due pigeonhole principle.

**Comment by BaurzhanSakhariev** [comment]
>Update on issue 2 (CrateDB only falsely says that HashJoin is not possible and uses NestedLoopJoin). However, result is correct so probably not worth fixing.

Hi @matriv can you please take a look? Do you agree with OR case? Motivation - don't overcomplicate code for the single use case where NestedLookJoin is better anyway. I think it's better for booleans in general but since we support already some cases no reason to change it but specifically changing not worth it?

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java [hunk 3]
```diff
@@ -119,4 +119,25 @@ public void test_equality_expression_followed_by_case_expression() {
...
+        assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isTrue();
+
+        // Compare column with constant
+        joinCondition = sqlExpressions.asSymbol("(t1.a >= 1) = (t2.b < 10)");
+        assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isTrue();
+
+        // Compare with column AND compare with constant.
+        joinCondition = sqlExpressions.asSymbol("(t1.a >= 1) = (t2.b <= t2.b AND t2.b < 10)");
+        assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isTrue();
+    }
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:142]: All those work on master.  Added to improve coverage (as initially I thought <= is the problem).

File: docs/appendices/release-notes/5.7.3.rst [hunk 4]
```diff
@@ -69,3 +69,11 @@ Fixes
 - Changed :ref:`scalar-pg_get_function_result` functions to be registered under
   :ref:`postgres-pg_catalog` schema, in order to be compatible with PostgreSQL
   behaviour.
+
+- Fixed an issue leading to incorrect results when joining tables on a boolean
+  expression, containing ``=`` operator and multiple relations as a part of the
```
Comment (matriv) [inline docs/appendices/release-notes/5.7.3.rst:74]: Rephrasing suggestion:

```
when joining tables with a ``=`` and on either side (or both) of the equality there is another ``=`` operator....

File: server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java [hunk 5]
```diff
@@ -90,18 +95,30 @@ public Void visitFunction(Function function, Context context) {
...
-
                     var right = arguments.get(1);
-                    var rightContext = new Context();
-                    right.accept(this, rightContext);
-                    if (leftContext.relations.size() == 1 &&
-                        rightContext.relations.size() == 1 &&
-                        !leftContext.relations.equals(rightContext.relations)) {
-                        context.isHashJoinPossible = true;
+                    if (context.insideEqualOperand) {
+                        // EQ operator inside either side of the JOIN condition.
```
Comment (matriv) [inline server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java:101]:
```suggestion
                        // EQ operator inside either side of the top level EQ operator```

File: server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java [hunk 6]
```diff
@@ -90,18 +95,30 @@ public Void visitFunction(Function function, Context context) {
...
+                        // We need to re-use the same context to ensure that we keep counting all relations of each JOIN side.
+                        left.accept(this, context);
+                        right.accept(this, context);
+                    } else {
+                        // EQ operator of the JOIN condition.
```
Comment (matriv) [inline server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java:106]:
```suggestion
                        // Top level EQ operator of the JOIN condition.
```

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java [hunk 7]
```diff
@@ -119,4 +119,25 @@ public void test_equality_expression_followed_by_case_expression() {
         Symbol joinCondition = sqlExpressions.asSymbol("t1.a = t1.a AND CASE 1 WHEN t1.a THEN false ELSE t2.b in (t2.b) END");
         assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isFalse();
     }
+
+    @Test
+    public void equality_and_many_relations_in_boolean_join_condition_hash_join_not_possible() {
```
Comment (matriv) [inline server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:124]:
```suggestion
    public void test_equality_and_many_relations_in_boolean_join_condition_hash_join_not_possible() {
```

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java [hunk 8]
```diff
@@ -119,4 +119,25 @@ public void test_equality_expression_followed_by_case_expression() {
...
+        Symbol joinCondition = sqlExpressions.asSymbol("(t1.a >= 1) = ((t1.a = t1.a) AND (t2.b <= t2.b))");
```
Comment (matriv) [inline server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:125]: maybe for future proof add on more test with more levels of nested `=`, e.g.:
```
expr1 = expr2 AND (expr3 = (t1.a = t1.a + 10)  AND (expr 5 = t1.b = t1.b - 5)
```

**Review by matriv — COMMENTED** [review]
Thx for the thorough investigation.
Looks good, left some comments/suggestions.

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java (hunk 3 above)
Comment (matriv) [reply server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:142]: +1

File: docs/appendices/release-notes/5.7.3.rst (hunk 4 above)
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/5.7.3.rst]: Thx, applied. BTW, I kept 'multiple relations involved" as issue is in combination of  `=` and many relations

Comment (matriv) [reply docs/appendices/release-notes/5.7.3.rst]: thx!

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java [hunk 9]
```diff
@@ -119,4 +119,25 @@ public void test_equality_expression_followed_by_case_expression() {
...
     }
+
+    @Test
+    public void test_equality_and_many_relations_in_boolean_join_condition_hash_join_not_possible() {
+        Symbol joinCondition = sqlExpressions.asSymbol("(t1.a >= 1) = ((t1.a = t1.a) AND (t2.b <= t2.b))");
+        assertThat(EquiJoinDetector.isHashJoinPossible(JoinType.INNER, joinCondition)).isFalse();
+    }
+
+    @Test
+    public void inequality_in_boolean_join_condition_hash_join_possible() {
```
Comment (matriv) [inline server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:130]: apologies, also here:
```suggestion
    public void test_inequality_in_boolean_join_condition_hash_join_possible() {
```

File: server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java (hunk 8 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:125]: mismatched input '<' expecting {<EOF>, 'AT', 'OR', 'AND', '+', '-', '*', '^', '/', '%', '||', '::', '.', '[', '&', '|', '#'}

Tried different versions, having this exception. Will check, probably we cannot even do nested nested equal.

Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:125]: Ok, had to wrap each side in extra parenthesis. Same in PG

Comment (matriv) [reply server/src/test/java/io/crate/planner/operators/EquiJoinDetectorTest.java:125]: yep, sorry, I wrote this in a hurry, thx for figuring out.

**Review by matriv — APPROVED** [review]
LGTM, thank you!

**Comment by BaurzhanSakhariev** [comment]
> Comparisons referencing the same column such as `(t1.a = t1.a)` or `(t2.b <= t2.b)` can be simplified to `true`. Instead of handling this inside the HashJoin-Detection we could introduce a step to simplify the join condition before.

We can simplify but we can have another query with the same problem: `EquiJoinDetector` not seeing multiple relations on  a JOIN side.

Example: `SELECT * FROM t0, t1 WHERE (t1.c1 >= 1)=((t1.c1 = 5) AND (t0.c1 <= 5))`
 
Here we don't use self comparison but without this change it will be false positive due to `EquiJoinDetector` missing that t0/t1 are on the same side.

**Comment by mkleen** [comment]
`(t1.c1 >= 1)=((t1.c1 = 5) AND (t0.c1 <= 5))` can be simplified to `(true = (t0.c1 = 5))`.

**Comment by mkleen** [comment]
Let us improve to simplify the join conditions instead of handling this all here.

**Pull request merged**